### PR TITLE
[yaml2obj][XOFF] Update yaml2obj for XCOFF to create valid XCOFF files in more cases.

### PR DESCRIPTION
**Pull request by stephenpeckham**
yaml2obj creates invalid object files even when the input was created by obj2yaml using a valid object file. On the other hand, yaml2obj is used to intentionally create invalid object files for testing purposes.

This update balances using specified input values when provided and computing file offsets and sizes if necessary.

### CODE REVIEW

**Review by jh7370 — COMMENTED** [review]
I don't think all the error cases have tests? Also you need to ensure that all your optional fields are tested (showing default behaviour when not specified).

File: llvm/lib/ObjectYAML/XCOFFEmitter.cpp [hunk 1]
```diff
@@ -40,6 +40,8 @@ class XCOFFWriter {
   bool writeXCOFF();
 
 private:
+  void ErrOverwrite(uint64_t currentOffset, uint64_t specifiedOffset,
```
Comment (jh7370) [inline llvm/lib/ObjectYAML/XCOFFEmitter.cpp:43]: Naming style (casing and this isn't a "verb phrase")

File: llvm/lib/ObjectYAML/XCOFFEmitter.cpp [hunk 2]
```diff
@@ -90,6 +92,13 @@ static void writeName(StringRef StrName, support::endian::Writer W) {
   W.write(NameRef);
 }
 
+void XCOFFWriter::ErrOverwrite(uint64_t CurrentOffset, uint64_t specifiedOffset,
+                               const Twine &fieldName) {
```
Comment (jh7370) [inline llvm/lib/ObjectYAML/XCOFFEmitter.cpp:96]: Coding style.

File: llvm/test/tools/yaml2obj/XCOFF/offset-check.yaml [hunk 3]
```diff
@@ -0,0 +1,156 @@
+## Case 1: an error is reported when FileOffsetToData is too small
+## RUN: not yaml2obj --docnum=1 %s -o %t 2>&1 | \
```
Comment (jh7370) [inline llvm/test/tools/yaml2obj/XCOFF/offset-check.yaml:2]: `##` is for comments. `#` for lit/FileCheck directives.

File: llvm/test/tools/yaml2obj/XCOFF/offset-check.yaml [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+## RUN:   FileCheck %s --check-prefix=ERROR1
+#
+# ERROR1: error: current file offset (60) is bigger than the specified FileOffsetToData for the .text section (48)
```
Comment (jh7370) [inline llvm/test/tools/yaml2obj/XCOFF/offset-check.yaml:5]: Nit: Include a blank line between your check and the YAML for improved readability. Applies throughout.

File: llvm/test/tools/yaml2obj/XCOFF/offset-check.yaml [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+    StorageClass:    C_EXT
+    NumberOfAuxEntries: 1
+    AuxEntries:
+      - Type:            AUX_CSECT
+        SymbolAlignmentAndType: 17
+        StorageMappingClass: XMC_PR
+        SectionOrLength: 8
+StringTable:     {}
+...
+## Case 2: an error is reported when FileOffsetToRelocations is too small
```
Comment (jh7370) [inline llvm/test/tools/yaml2obj/XCOFF/offset-check.yaml:53]: Nit: Include a blank line after the YAML too.

File: llvm/test/tools/yaml2obj/XCOFF/offset-check.yaml [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+    StorageClass:    C_EXT
+    NumberOfAuxEntries: 1
+    AuxEntries:
+      - Type:            AUX_CSECT
+        SymbolAlignmentAndType: 17
+        StorageMappingClass: XMC_PR
+        SectionOrLength: 8
+StringTable:     {}
+...
+## Case 1: an error is reported when the OffsetToSymbolTable is too small
```
Comment (jh7370) [inline llvm/test/tools/yaml2obj/XCOFF/offset-check.yaml:105]: This isn't "Case 1"...

File: llvm/lib/ObjectYAML/XCOFFEmitter.cpp [hunk 7]
```diff
@@ -98,30 +107,42 @@ bool XCOFFWriter::nameShouldBeInStringTable(StringRef SymbolName) {
...
+        CurrentOffset = InitSection.FileOffsetToRelocations;
+      } else
+        InitSection.FileOffsetToRelocations = CurrentOffset;
+      CurrentOffset += UsedSize;
       if (CurrentOffset > MaxRawDataSize) {
-        ErrHandler("maximum object size of" + Twine(MaxRawDataSize) +
-                   "exceeded when writing relocation data");
+        ErrHandler("maximum object size (" + Twine(MaxRawDataSize) +
+                   ") exceeded when writing relocation data for section " +
+                   Twine(InitSection.SectionName));
```
Comment (jh7370) [inline llvm/lib/ObjectYAML/XCOFFEmitter.cpp:134]: Is this tested somewhere? If so, could that test be expanded to check the section name that is now in the message? If not, it should be tested.

File: llvm/lib/ObjectYAML/XCOFFEmitter.cpp [hunk 8]
```diff
@@ -134,23 +155,58 @@ bool XCOFFWriter::initSectionHeader(uint64_t &CurrentOffset) {
...
+        CurrentOffset = alignTo(CurrentOffset, DefaultSectionAlign);
+        InitSections[I].FileOffsetToData = CurrentOffset;
+      }
       CurrentOffset += InitSections[I].SectionData.binary_size();
-      // Ensure the offset is aligned to DefaultSectionAlign.
-      CurrentOffset = alignTo(CurrentOffset, DefaultSectionAlign);
-      InitSections[I].Size = CurrentOffset - InitSections[I].FileOffsetToData;
-      CurrentSecAddr += InitSections[I].Size;
+      if (CurrentOffset > MaxRawDataSize) {
+        ErrHandler("maximum object size (" + Twine(MaxRawDataSize) +
```
Comment (jh7370) [inline llvm/lib/ObjectYAML/XCOFFEmitter.cpp:205]: Is this tested?

File: llvm/lib/ObjectYAML/XCOFFEmitter.cpp [hunk 9]
```diff
@@ -254,12 +310,20 @@ bool XCOFFWriter::initFileHeader(uint64_t CurrentOffset) {
...
+      }
+      CurrentOffset = Obj.Header.SymbolTableOffset;
+    }
     InitFileHdr.SymbolTableOffset = CurrentOffset;
     CurrentOffset +=
         InitFileHdr.NumberOfSymTableEntries * XCOFF::SymbolTableEntrySize;
     if (CurrentOffset > MaxRawDataSize) {
-      ErrHandler("maximum object size of" + Twine(MaxRawDataSize) +
-                 "exceeded when writing symbols");
+      ErrHandler("maximum object size of " + Twine(MaxRawDataSize) +
```
Comment (jh7370) [inline llvm/lib/ObjectYAML/XCOFFEmitter.cpp:325]: Is this tested?

File: llvm/lib/ObjectYAML/XCOFFEmitter.cpp [hunk 10]
```diff
@@ -486,10 +543,10 @@ bool XCOFFWriter::writeSectionData() {
     XCOFFYAML::Section YamlSec = Obj.Sections[I];
     if (YamlSec.SectionData.binary_size()) {
       // Fill the padding size with zeros.
-      int64_t PaddingSize =
-          InitSections[I].FileOffsetToData - (W.OS.tell() - StartOffset);
+      int64_t PaddingSize = (uint64_t)InitSections[I].FileOffsetToData -
+                            (W.OS.tell() - StartOffset);
       if (PaddingSize < 0) {
-        ErrHandler("redundant data was written before section data");
+        ErrHandler("unexpected data was written before section data");
```
Comment (jh7370) [inline llvm/lib/ObjectYAML/XCOFFEmitter.cpp:549]: Test case?

File: llvm/lib/ObjectYAML/XCOFFEmitter.cpp [hunk 11]
```diff
@@ -506,7 +563,7 @@ bool XCOFFWriter::writeRelocations() {
       int64_t PaddingSize =
           InitSections[I].FileOffsetToRelocations - (W.OS.tell() - StartOffset);
       if (PaddingSize < 0) {
-        ErrHandler("redundant data was written before relocations");
+        ErrHandler("unexpected data was written before relocations");
```
Comment (jh7370) [inline llvm/lib/ObjectYAML/XCOFFEmitter.cpp:566]: Test case?

File: llvm/lib/ObjectYAML/XCOFFEmitter.cpp [hunk 12]
```diff
@@ -645,9 +702,9 @@ void XCOFFWriter::writeAuxSymbol(
 
 bool XCOFFWriter::writeSymbols() {
   int64_t PaddingSize =
-      (uint64_t)InitFileHdr.SymbolTableOffset - (W.OS.tell() - StartOffset);
+      InitFileHdr.SymbolTableOffset - (W.OS.tell() - StartOffset);
   if (PaddingSize < 0) {
-    ErrHandler("redundant data was written before symbols");
+    ErrHandler("unexpected data was written before symbols");
```
Comment (jh7370) [inline llvm/lib/ObjectYAML/XCOFFEmitter.cpp:707]: Test case?

**Review by jh7370 — COMMENTED** [review]
> Also you need to ensure that all your optional fields are tested (showing default behaviour when not specified).

Have you addressed this comment?

File: llvm/lib/ObjectYAML/XCOFFEmitter.cpp (hunk 10 above)
Comment (stephenpeckham) [reply llvm/lib/ObjectYAML/XCOFFEmitter.cpp]: This error (and the 2 similar ones) should be asserts. There's no need to change the wording of the message.

**Comment by stephenpeckham** [comment]
> > Also you need to ensure that all your optional fields are tested (showing default behaviour when not specified).
> 
> Have you addressed this comment?
Adding tests for every optional field is beyond the scope of this PR.

**Review by jh7370 — COMMENTED** [review]
I haven't attempted to trace things fully, but the `AuxHeaderSize` field is now a `std::optional` in this patch. Is that not something that is exposed to the user?

**Comment by stephenpeckham** [comment]
> I haven't attempted to trace things fully, but the `AuxHeaderSize` field is now a `std::optional` in this patch. Is that not something that is exposed to the user?

I thought I had a reason to make AuxHeaderSize optional , but I it looks like that specific change isn't necessary, at least for now.  I'm not sure what you mean by exposing an optional field to the user.  In any case, the XCOFF auxiliary header support needs some additional changes (in a future PR--the auxiliary fields can be read by yaml2obj, but are not currently printed by obj2yaml.

**Review by jh7370 — APPROVED** [review]
> > > I haven't attempted to trace things fully, but the `AuxHeaderSize` field is now a `std::optional` in this patch. Is that not something that is exposed to the user?
> 
> I thought I had a reason to make AuxHeaderSize optional , but I it looks like that specific change isn't necessary, at least for now. I'm not sure what you mean by exposing an optional field to the user. In any case, the XCOFF auxiliary header support needs some additional changes (in a future PR--the auxiliary fields can be read by yaml2obj, but are not currently printed by obj2yaml.

I don't know why, but I had in my head that you'd changed code that might mean that people could omit a field that they couldn't before, in which case that should be tested. Looking again, I don't think there is, sorry for the fuss and LGTM.

**Pull request merged**